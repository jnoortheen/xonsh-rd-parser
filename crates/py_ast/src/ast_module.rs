/// A wrapper around the Python ast module.
use pyo3::prelude::PyModule;
use pyo3::types::{IntoPyDict, PyAnyMethods};
use pyo3::{IntoPyObject, IntoPyObjectExt, PyObject, PyResult, Python};
use ruff_source_file::SourceCode;
use ruff_text_size::TextRange;

pub struct AstModule<'py> {
    obj: PyObject,
    pub py: Python<'py>,
    source_code: &'py SourceCode<'py, 'py>,
}

impl<'py> AstModule<'py> {
    pub fn new(py: Python<'py>, source_code: &'py SourceCode) -> PyResult<Self> {
        let obj: PyObject = PyModule::import(py, "ast")?.unbind().into();
        Ok(Self {
            obj,
            py,
            source_code,
        })
    }
    pub(crate) fn attr(&self, name: &str) -> PyResult<Self> {
        let obj = self.obj.getattr(self.py, name)?;
        Ok(AstModule {
            obj,
            py: self.py,
            source_code: self.source_code,
        })
    }
    pub fn call<T: IntoPyDict<'py>>(&self, range: TextRange, kwargs: T) -> PyResult<PyObject> {
        let dict = kwargs.into_py_dict(self.py)?;
        for (key, value) in self.location(range) {
            dict.set_item(key, value)?;
        }
        Ok(self.obj.bind(self.py).call((), Some(&dict))?.into())
    }
    pub(crate) fn location(&self, range: TextRange) -> [(&'static str, usize); 4] {
        let start = self.source_code.source_location(range.start());
        let end = self.source_code.source_location(range.end());

        [
            ("lineno", start.row.get()),
            ("col_offset", start.column.get()),
            ("end_lineno", end.row.get()),
            ("end_col_offset", end.column.get()),
        ]
    }
    pub fn to_const<T: IntoPyObject<'py>>(&self, range: TextRange, value: T) -> PyResult<PyObject> {
        self.attr("Constant")?.call(range, [("value", value)])
    }
    pub fn call0(&self) -> PyResult<PyObject> {
        Ok(self.obj.bind(self.py).call0()?.into())
    }
    pub fn callk<T: IntoPyDict<'py>>(&self, kwargs: T) -> PyResult<PyObject> {
        let kwargs = kwargs.into_py_dict(self.py)?;
        Ok(self.obj.bind(self.py).call((), Some(&kwargs))?.into())
    }

    pub fn empty_list(&self) -> PyResult<PyObject> {
        let empty_vec: Vec<i32> = vec![]; // Explicitly specify the type of Vec
        empty_vec.into_py_any(self.py)
    }
}
