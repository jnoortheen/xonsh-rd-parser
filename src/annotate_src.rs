use ruff_python_parser::{ParseError, ParseErrorType};
use ruff_source_file::{LineIndex, OneIndexed, SourceCode};
use ruff_text_size::TextRange;
use std::fmt::Formatter;

use annotate_snippets::display_list::{DisplayList, FormatOptions};
use annotate_snippets::snippet::{AnnotationType, Slice, Snippet, SourceAnnotation};

pub(crate) fn to_exc_msg(src: &str, filename: &str, err: &ParseError) -> String {
    let line_index = LineIndex::from_source_text(src);
    let source_code = SourceCode::new(src, &line_index);
    let code_frame = CodeFrame::new(&source_code, err);
    format!("{err} in {filename}:\n{code_frame}")
}
pub(crate) struct CodeFrame<'a> {
    range: TextRange,
    error: &'a ParseErrorType,
    source: &'a SourceCode<'a, 'a>,
}

impl<'a> CodeFrame<'a> {
    pub(crate) fn new(source: &'a SourceCode<'_, '_>, error: &'a ParseError) -> Self {
        CodeFrame {
            range: error.location,
            error: &error.error,
            source,
        }
    }
}

impl std::fmt::Display for CodeFrame<'_> {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        // Copied and modified from ruff_linter/src/message/text.rs
        let content_start_index = self.source.line_index(self.range.start());
        let mut start_index = content_start_index.saturating_sub(2);

        // Trim leading empty lines.
        while start_index < content_start_index {
            if !self.source.line_text(start_index).trim().is_empty() {
                break;
            }
            start_index = start_index.saturating_add(1);
        }

        let content_end_index = self.source.line_index(self.range.end());
        let mut end_index = content_end_index
            .saturating_add(2)
            .min(OneIndexed::from_zero_indexed(self.source.line_count()));

        // Trim trailing empty lines.
        while end_index > content_end_index {
            if !self.source.line_text(end_index).trim().is_empty() {
                break;
            }

            end_index = end_index.saturating_sub(1);
        }

        let start_offset = self.source.line_start(start_index);
        let end_offset = self.source.line_end(end_index);

        let annotation_range = self.range - start_offset;
        let source = self.source.slice(TextRange::new(start_offset, end_offset));

        let start_char = source[TextRange::up_to(annotation_range.start())]
            .chars()
            .count();

        let char_length = source[annotation_range].chars().count();
        let label = format!("Syntax Error: {error}", error = self.error);

        let snippet = Snippet {
            title: None,
            slices: vec![Slice {
                source,
                line_start: start_index.get(),
                annotations: vec![SourceAnnotation {
                    label: &label,
                    annotation_type: AnnotationType::Error,
                    range: (start_char, start_char + char_length),
                }],
                // The origin (file name, line number, and column number) is already encoded
                // in the `label`.
                origin: None,
                fold: false,
            }],
            footer: Vec::new(),
            opt: FormatOptions::default(),
        };

        writeln!(f, "{message}", message = DisplayList::from(snippet))
    }
}
