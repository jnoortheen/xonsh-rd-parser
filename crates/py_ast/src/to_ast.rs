/// Convert Ruff's AST to Python AST.
use crate::ast_module::AstModule;
use pyo3::{IntoPy, PyObject};

type PyResult = pyo3::PyResult<PyObject>;

pub trait ToAst {
    fn to_ast(&self, module: &AstModule) -> PyResult;
}

impl<T: ToAst> ToAst for Option<T> {
    fn to_ast(&self, module: &AstModule) -> PyResult {
        match self {
            Some(param) => param.to_ast(module),
            None => Ok(module.py.None()),
        }
    }
}
impl<T: ToAst> ToAst for Box<T> {
    fn to_ast(&self, module: &AstModule) -> PyResult {
        self.as_ref().to_ast(module)
    }
}
impl<T: ToAst> ToAst for Vec<T> {
    fn to_ast(&self, module: &AstModule) -> PyResult {
        let mut body: Vec<PyObject> = vec![];
        for stmt in self {
            body.push(stmt.to_ast(module)?);
        }
        Ok(body.into_py(module.py))
    }
}
impl<T: ToAst> ToAst for [T] {
    fn to_ast(&self, module: &AstModule) -> PyResult {
        let mut body: Vec<PyObject> = vec![];
        for stmt in self {
            body.push(stmt.to_ast(module)?);
        }
        Ok(body.into_py(module.py))
    }
}
