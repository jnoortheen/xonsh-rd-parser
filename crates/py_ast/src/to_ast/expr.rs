use std::borrow::Cow;
use std::vec;

use super::ToAst;
use crate::ast_module::AstModule;
use num_complex::Complex;
use pyo3::{IntoPyObjectExt, PyObject};
use ruff_python_ast::str_prefix::StringLiteralPrefix;
use ruff_python_ast::*;
use ruff_text_size::Ranged;

type PyResult = pyo3::PyResult<PyObject>;

impl ToAst for Expr {
    fn to_ast(&self, module: &AstModule) -> PyResult {
        match self {
            Expr::BoolOp(expr) => expr.to_ast(module),
            Expr::Named(expr) => expr.to_ast(module),
            Expr::BinOp(expr) => expr.to_ast(module),
            Expr::UnaryOp(expr) => expr.to_ast(module),
            Expr::Lambda(expr) => expr.to_ast(module),
            Expr::If(expr) => expr.to_ast(module),
            Expr::Dict(expr) => expr.to_ast(module),
            Expr::Set(expr) => expr.to_ast(module),
            Expr::List(expr) => expr.to_ast(module),
            Expr::ListComp(expr) => expr.to_ast(module),
            Expr::SetComp(expr) => expr.to_ast(module),
            Expr::DictComp(expr) => expr.to_ast(module),
            Expr::Generator(expr) => expr.to_ast(module),
            Expr::Tuple(expr) => expr.to_ast(module),
            Expr::Slice(expr) => expr.to_ast(module),
            Expr::Call(expr) => expr.to_ast(module),
            Expr::Attribute(expr) => expr.to_ast(module),
            Expr::Subscript(expr) => expr.to_ast(module),
            Expr::Starred(expr) => expr.to_ast(module),
            Expr::Name(expr) => expr.to_ast(module),
            Expr::Await(expr) => expr.to_ast(module),
            Expr::Yield(expr) => expr.to_ast(module),
            Expr::YieldFrom(expr) => expr.to_ast(module),
            Expr::Compare(expr) => expr.to_ast(module),
            Expr::FString(expr) => expr.to_ast(module),
            Expr::StringLiteral(expr) => expr.to_ast(module),
            Expr::BytesLiteral(expr) => expr.to_ast(module),
            Expr::NumberLiteral(expr) => expr.to_ast(module),
            Expr::BooleanLiteral(expr) => expr.to_ast(module),
            Expr::NoneLiteral(expr) => expr.to_ast(module),
            Expr::EllipsisLiteral(expr) => expr.to_ast(module),
            Expr::IpyEscapeCommand(_expr) => unreachable!(),
        }
    }
}
impl ToAst for ExprNumberLiteral {
    fn to_ast(&self, module: &AstModule) -> PyResult {
        let value = match &self.value {
            Number::Int(value) => value.as_u64().into_py_any(module.py),
            Number::Float(value) => value.into_py_any(module.py),
            Number::Complex { real, imag } => Complex {
                re: *real,
                im: *imag,
            }
            .into_py_any(module.py),
        };
        module.to_const(self.range(), value?.into_py_any(module.py)?)
    }
}
impl ToAst for ExprEllipsisLiteral {
    fn to_ast(&self, module: &AstModule) -> PyResult {
        module.to_const(self.range(), module.py.Ellipsis())
    }
}
impl ToAst for ExprNoneLiteral {
    fn to_ast(&self, module: &AstModule) -> PyResult {
        module.to_const(self.range(), module.py.None())
    }
}
impl ToAst for ExprBooleanLiteral {
    fn to_ast(&self, module: &AstModule) -> PyResult {
        module.to_const(self.range(), self.value)
    }
}
impl ToAst for ExprBytesLiteral {
    fn to_ast(&self, module: &AstModule) -> PyResult {
        let value: Cow<[u8]> = Cow::Owned(self.value.bytes().collect::<Vec<u8>>());
        module.to_const(self.range(), value)
    }
}
impl ToAst for ExprStringLiteral {
    fn to_ast(&self, module: &AstModule) -> PyResult {
        let kind = if self.value.is_unicode() {
            Some("u")
        } else {
            None
        };
        let kwargs = [
            (
                "value",
                self.value.to_str().to_string().into_py_any(module.py)?,
            ),
            ("kind", kind.into_py_any(module.py)?),
        ];
        module.attr("Constant")?.call(self.range(), kwargs)
    }
}
impl ToAst for ExprFString {
    fn to_ast(&self, module: &AstModule) -> PyResult {
        let mut parts = vec![];
        for p in self.value.as_slice().iter() {
            match p {
                FStringPart::Literal(s) => {
                    let kind: PyObject = match s.flags.prefix() {
                        StringLiteralPrefix::Unicode => Some("u"),
                        _ => None,
                    }
                    .into_py_any(module.py)?;
                    let obj = module.attr("Constant")?.call(
                        self.range(),
                        [
                            ("kind", kind),
                            ("value", s.as_str().into_py_any(module.py)?),
                        ],
                    )?;
                    parts.push(obj);
                }
                FStringPart::FString(fs) => {
                    for p in fs.elements.iter() {
                        parts.push(p.to_ast(module)?);
                    }
                }
            }
        }
        module
            .attr("JoinedStr")?
            .call(self.range, [("values", parts)])
    }
}
impl ToAst for ConversionFlag {
    fn to_ast(&self, module: &AstModule) -> PyResult {
        let flag = *self as i8;
        flag.into_py_any(module.py)
    }
}
impl ToAst for FStringFormatSpec {
    fn to_ast(&self, module: &AstModule) -> PyResult {
        let mut values = vec![];
        for value in self.elements.iter() {
            values.push(value.to_ast(module)?);
        }
        module
            .attr("JoinedStr")?
            .call(self.range(), [("values", values.into_py_any(module.py)?)])
    }
}
impl ToAst for FStringExpressionElement {
    fn to_ast(&self, module: &AstModule) -> PyResult {
        module.attr("FormattedValue")?.call(
            self.range,
            [
                ("value", self.expression.to_ast(module)?),
                ("conversion", self.conversion.to_ast(module)?),
                ("format_spec", self.format_spec.to_ast(module)?),
            ],
        )
    }
}
impl ToAst for FStringElement {
    fn to_ast(&self, module: &AstModule) -> PyResult {
        let obj = match self {
            FStringElement::Literal(literal) => module.to_const(
                self.range(),
                literal.value.to_string().into_py_any(module.py)?,
            )?,
            FStringElement::Expression(expr) => expr.to_ast(module)?,
        };
        Ok(obj)
    }
}
impl ToAst for CmpOp {
    fn to_ast(&self, module: &AstModule) -> PyResult {
        let obj = match self {
            CmpOp::Eq => module.attr("Eq")?.call0()?.into(),
            CmpOp::NotEq => module.attr("NotEq")?.call0()?.into(),
            CmpOp::Lt => module.attr("Lt")?.call0()?.into(),
            CmpOp::LtE => module.attr("LtE")?.call0()?.into(),
            CmpOp::Gt => module.attr("Gt")?.call0()?.into(),
            CmpOp::GtE => module.attr("GtE")?.call0()?.into(),
            CmpOp::Is => module.attr("Is")?.call0()?.into(),
            CmpOp::IsNot => module.attr("IsNot")?.call0()?.into(),
            CmpOp::In => module.attr("In")?.call0()?.into(),
            CmpOp::NotIn => module.attr("NotIn")?.call0()?.into(),
        };
        Ok(obj)
    }
}
impl ToAst for ExprCompare {
    fn to_ast(&self, module: &AstModule) -> PyResult {
        module.attr("Compare")?.call(
            self.range,
            [
                ("left", self.left.to_ast(module)?),
                ("ops", self.ops.to_ast(module)?),
                ("comparators", self.comparators.to_ast(module)?),
            ],
        )
    }
}
impl ToAst for ExprYieldFrom {
    fn to_ast(&self, module: &AstModule) -> PyResult {
        module
            .attr("YieldFrom")?
            .call(self.range, [("value", self.value.to_ast(module)?)])
    }
}
impl ToAst for ExprYield {
    fn to_ast(&self, module: &AstModule) -> PyResult {
        module
            .attr("Yield")?
            .call(self.range, [("value", self.value.to_ast(module)?)])
    }
}
impl ToAst for ExprAwait {
    fn to_ast(&self, module: &AstModule) -> PyResult {
        module
            .attr("Await")?
            .call(self.range, [("value", self.value.to_ast(module)?)])
    }
}
impl ToAst for ExprName {
    fn to_ast(&self, module: &AstModule) -> PyResult {
        module.attr("Name")?.call(
            self.range,
            [
                ("id", self.id.to_ast(module)?),
                ("ctx", self.ctx.to_ast(module)?),
            ],
        )
    }
}
impl ToAst for name::Name {
    fn to_ast(&self, module: &AstModule) -> PyResult {
        self.as_str().to_string().into_py_any(module.py)
    }
}
impl ToAst for ExprStarred {
    fn to_ast(&self, module: &AstModule) -> PyResult {
        module.attr("Starred")?.call(
            self.range,
            [
                ("value", self.value.to_ast(module)?),
                ("ctx", self.ctx.to_ast(module)?),
            ],
        )
    }
}
impl ToAst for ExprSubscript {
    fn to_ast(&self, module: &AstModule) -> PyResult {
        module.attr("Subscript")?.call(
            self.range,
            [
                ("ctx", self.ctx.to_ast(module)?),
                ("value", self.value.to_ast(module)?),
                ("slice", self.slice.to_ast(module)?),
            ],
        )
    }
}
impl ToAst for ExprAttribute {
    fn to_ast(&self, module: &AstModule) -> PyResult {
        module.attr("Attribute")?.call(
            self.range,
            [
                ("value", self.value.to_ast(module)?),
                ("attr", self.attr.to_ast(module)?),
                ("ctx", self.ctx.to_ast(module)?),
            ],
        )
    }
}
impl ToAst for Keyword {
    fn to_ast(&self, module: &AstModule) -> PyResult {
        module.attr("keyword")?.call(
            self.range,
            [
                ("arg", self.arg.to_ast(module)?),
                ("value", self.value.to_ast(module)?),
            ],
        )
    }
}
impl ToAst for ExprCall {
    fn to_ast(&self, module: &AstModule) -> PyResult {
        module.attr("Call")?.call(
            self.range,
            [
                ("func", self.func.to_ast(module)?),
                ("args", self.arguments.args.to_ast(module)?),
                ("keywords", self.arguments.keywords.to_ast(module)?),
            ],
        )
    }
}
impl ToAst for ExprSlice {
    fn to_ast(&self, module: &AstModule) -> PyResult {
        module.attr("Slice")?.call(
            self.range,
            [
                ("lower", self.lower.to_ast(module)?),
                ("upper", self.upper.to_ast(module)?),
                ("step", self.step.to_ast(module)?),
            ],
        )
    }
}
impl ToAst for ExprContext {
    fn to_ast(&self, module: &AstModule) -> PyResult {
        let obj = match self {
            ExprContext::Del => module.attr("Del")?.call0()?.into(),
            ExprContext::Load => module.attr("Load")?.call0()?.into(),
            ExprContext::Store => module.attr("Store")?.call0()?.into(),
            ExprContext::Invalid => todo!(),
        };
        Ok(obj)
    }
}
impl ToAst for ExprTuple {
    fn to_ast(&self, module: &AstModule) -> PyResult {
        module.attr("Tuple")?.call(
            self.range,
            [
                ("elts", self.elts.to_ast(module)?),
                ("ctx", self.ctx.to_ast(module)?),
            ],
        )
    }
}
impl ToAst for ExprGenerator {
    fn to_ast(&self, module: &AstModule) -> PyResult {
        module.attr("GeneratorExp")?.call(
            self.range,
            [
                ("elt", self.elt.to_ast(module)?),
                ("generators", self.generators.to_ast(module)?),
            ],
        )
    }
}
impl ToAst for Comprehension {
    fn to_ast(&self, module: &AstModule) -> PyResult {
        module.attr("comprehension")?.callk([
            ("target", self.target.to_ast(module)?),
            ("iter", self.iter.to_ast(module)?),
            ("ifs", self.ifs.to_ast(module)?),
            ("is_async", (self.is_async as u8).into_py_any(module.py)?),
        ])
    }
}
impl ToAst for ExprDictComp {
    fn to_ast(&self, module: &AstModule) -> PyResult {
        module.attr("DictComp")?.call(
            self.range,
            [
                ("key", self.key.to_ast(module)?),
                ("value", self.value.to_ast(module)?),
                ("generators", self.generators.to_ast(module)?),
            ],
        )
    }
}
impl ToAst for ExprSetComp {
    fn to_ast(&self, module: &AstModule) -> PyResult {
        module.attr("SetComp")?.call(
            self.range,
            [
                ("elt", self.elt.to_ast(module)?),
                ("generators", self.generators.to_ast(module)?),
            ],
        )
    }
}
impl ToAst for ExprListComp {
    fn to_ast(&self, module: &AstModule) -> PyResult {
        module.attr("ListComp")?.call(
            self.range,
            [
                ("elt", self.elt.to_ast(module)?),
                ("generators", self.generators.to_ast(module)?),
            ],
        )
    }
}
impl ToAst for ExprList {
    fn to_ast(&self, module: &AstModule) -> PyResult {
        module.attr("List")?.call(
            self.range,
            [
                ("elts", self.elts.to_ast(module)?),
                ("ctx", self.ctx.to_ast(module)?),
            ],
        )
    }
}
impl ToAst for ExprSet {
    fn to_ast(&self, module: &AstModule) -> PyResult {
        module
            .attr("Set")?
            .call(self.range, [("elts", self.elts.to_ast(module)?)])
    }
}
impl ToAst for ExprDict {
    fn to_ast(&self, module: &AstModule) -> PyResult {
        let mut keys = vec![];
        let mut values = vec![];
        for item in self.items.iter() {
            keys.push(item.key.to_ast(module)?);
            values.push(item.value.to_ast(module)?);
        }
        module
            .attr("Dict")?
            .call(self.range, [("keys", keys), ("values", values)])
    }
}
impl ToAst for ExprIf {
    fn to_ast(&self, module: &AstModule) -> PyResult {
        module.attr("IfExp")?.call(
            self.range,
            [
                ("test", self.test.to_ast(module)?),
                ("body", self.body.to_ast(module)?),
                ("orelse", self.orelse.to_ast(module)?),
            ],
        )
    }
}

struct OptionalParameters(pub Option<Box<Parameters>>);

// special case for Parameters
impl ToAst for OptionalParameters {
    fn to_ast(&self, module: &AstModule) -> PyResult {
        match &self.0 {
            Some(parameters) => parameters.to_ast(module),
            None => module.attr("arguments")?.callk([
                ("posonlyargs", module.empty_list()?),
                ("args", module.empty_list()?),
                ("defaults", module.empty_list()?),
                ("kwonlyargs", module.empty_list()?),
                ("kw_defaults", module.empty_list()?),
                ("vararg", module.py.None()),
                ("kwarg", module.py.None()),
            ]),
        }
    }
}
impl ToAst for ExprLambda {
    fn to_ast(&self, module: &AstModule) -> PyResult {
        let params = OptionalParameters(self.parameters.clone());
        module.attr("Lambda")?.call(
            self.range,
            [
                ("args", params.to_ast(module)?),
                ("body", self.body.to_ast(module)?),
            ],
        )
    }
}
impl ToAst for ExprBinOp {
    fn to_ast(&self, module: &AstModule) -> PyResult {
        module.attr("BinOp")?.call(
            self.range,
            [
                ("left", self.left.to_ast(module)?),
                ("op", self.op.to_ast(module)?),
                ("right", self.right.to_ast(module)?),
            ],
        )
    }
}
impl ToAst for UnaryOp {
    fn to_ast(&self, module: &AstModule) -> PyResult {
        let obj = match self {
            UnaryOp::Invert => module.attr("Invert")?.call0()?.into(),
            UnaryOp::Not => module.attr("Not")?.call0()?.into(),
            UnaryOp::UAdd => module.attr("UAdd")?.call0()?.into(),
            UnaryOp::USub => module.attr("USub")?.call0()?.into(),
        };
        Ok(obj)
    }
}
impl ToAst for ExprUnaryOp {
    fn to_ast(&self, module: &AstModule) -> PyResult {
        module.attr("UnaryOp")?.call(
            self.range,
            [
                ("op", self.op.to_ast(module)?),
                ("operand", self.operand.to_ast(module)?),
            ],
        )
    }
}
impl ToAst for Operator {
    fn to_ast(&self, module: &AstModule) -> PyResult {
        let obj = match self {
            Operator::Add => module.attr("Add")?.call0()?.into(),
            Operator::Sub => module.attr("Sub")?.call0()?.into(),
            Operator::Mult => module.attr("Mult")?.call0()?.into(),
            Operator::MatMult => module.attr("MatMult")?.call0()?.into(),
            Operator::Div => module.attr("Div")?.call0()?.into(),
            Operator::Mod => module.attr("Mod")?.call0()?.into(),
            Operator::Pow => module.attr("Pow")?.call0()?.into(),
            Operator::LShift => module.attr("LShift")?.call0()?.into(),
            Operator::RShift => module.attr("RShift")?.call0()?.into(),
            Operator::BitOr => module.attr("BitOr")?.call0()?.into(),
            Operator::BitXor => module.attr("BitXor")?.call0()?.into(),
            Operator::FloorDiv => module.attr("FloorDiv")?.call0()?.into(),
            Operator::BitAnd => module.attr("BitAnd")?.call0()?.into(),
        };
        Ok(obj)
    }
}
impl ToAst for BoolOp {
    fn to_ast(&self, module: &AstModule) -> PyResult {
        let obj = match self {
            BoolOp::And | BoolOp::And2 => module.attr("And")?.call0()?.into(),
            BoolOp::Or | BoolOp::Or2 => module.attr("Or")?.call0()?.into(),
        };
        Ok(obj)
    }
}

impl ToAst for ExprNamed {
    fn to_ast(&self, module: &AstModule) -> PyResult {
        module.attr("NamedExpr")?.call(
            self.range,
            [
                ("target", self.target.to_ast(module)?),
                ("value", self.value.to_ast(module)?),
            ],
        )
    }
}

impl ToAst for ExprBoolOp {
    fn to_ast(&self, module: &AstModule) -> PyResult {
        module.attr("BoolOp")?.call(
            self.range,
            [
                ("op", self.op.to_ast(module)?),
                ("values", self.values.to_ast(module)?),
            ],
        )
    }
}
