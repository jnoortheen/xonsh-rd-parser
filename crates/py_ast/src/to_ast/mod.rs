mod expr;
mod r#match;
mod stmt;

use crate::ast_module::AstModule;
use pyo3::{IntoPyObjectExt, PyObject};

type PyResult = pyo3::PyResult<PyObject>;

pub trait ToAst {
    fn to_ast(&self, module: &AstModule) -> PyResult;
}
// speciailized implementations
impl<T: ToAst> ToAst for Option<T> {
    fn to_ast(&self, module: &AstModule) -> PyResult {
        match self {
            Some(param) => param.to_ast(module),
            None => Ok(module.py.None()),
        }
    }
}
impl<T: ToAst> ToAst for Box<T> {
    fn to_ast(&self, module: &AstModule) -> PyResult {
        self.as_ref().to_ast(module)
    }
}
// Shared implementation for sequences
fn to_ast_sequence<T: ToAst>(items: &[T], module: &AstModule) -> PyResult {
    let py_objects: Vec<PyObject> = items
        .iter()
        .map(|item| item.to_ast(module))
        .collect::<Result<_, _>>()?;
    Ok(py_objects.into_py_any(module.py)?)
}

impl<T: ToAst> ToAst for Vec<T> {
    fn to_ast(&self, module: &AstModule) -> PyResult {
        to_ast_sequence(self, module)
    }
}

impl<T: ToAst> ToAst for [T] {
    fn to_ast(&self, module: &AstModule) -> PyResult {
        to_ast_sequence(self, module)
    }
}
