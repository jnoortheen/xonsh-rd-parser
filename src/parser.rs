use py_ast::ast_module::AstModule;
use py_ast::to_ast::ToAst;
use pyo3::exceptions::PySyntaxError;
use pyo3::{PyObject, PyResult, Python};
use ruff_source_file::{LineIndex, SourceCode};

pub fn parse_str<'py>(
    py: Python<'py>,
    src: &'py str,
    filename: Option<&'py str>,
) -> PyResult<PyObject> {
    let parsed = ruff_python_parser::parse_module(src);
    match parsed {
        Ok(parsed) => {
            let line_index = LineIndex::from_source_text(src);
            let source_code = SourceCode::new(src, &line_index);
            let tree = parsed.into_syntax();
            let module = AstModule::new(py, &source_code)?;
            tree.to_ast(&module)
        }
        Err(err) => {
            let filename = filename.unwrap_or("<string>");
            let msg = crate::annotate_src::to_exc_msg(src, filename, &err);
            let err = PySyntaxError::new_err(msg);
            Err(err)
        }
    }
}

#[cfg(test)]
mod tests {
    use crate::annotate_src::CodeFrame;
    use ruff_python_parser::{parse_unchecked, Mode};
    use ruff_source_file::{LineIndex, SourceCode};
    use ruff_text_size::{Ranged, TextLen};
    use std::fmt::Write;

    fn test_valid_source<'a>(source: &'a str) -> String {
        let parsed = parse_unchecked(&source, Mode::Module);

        println!("Tokens: ");
        for token in parsed.tokens() {
            println!("{token:?} {}", source[token.range()].to_string());
        }
        println!("length: {:?}", source.text_len());

        if !parsed.is_valid() {
            let line_index = LineIndex::from_source_text(&source);
            let source_code = SourceCode::new(&source, &line_index);

            let mut message = "Expected no syntax errors for a valid \
            program but the parser generated the following errors:\n"
                .to_string();

            for error in parsed.errors() {
                let frame = CodeFrame::new(&source_code, error);
                writeln!(&mut message, "{frame}\n").unwrap();
                writeln!(&mut message,).unwrap();
            }

            panic!("{source:?}: {message}");
        }

        let mut output = String::new();
        writeln!(&mut output, "## AST").unwrap();
        writeln!(&mut output, "\n```\n{:#?}\n```", parsed.syntax()).unwrap();

        assert!(parsed.is_valid());
        output
    }

    #[test]
    fn test_tmp() {
        // let source = r#"![a@$(echo 1 2)b a]"#;
        let source = r#"!(ls @(None or "."))"#;
        // let source = r#"![a b c2]"#;
        // let source = r#"print('hello')"#;
        let _output = test_valid_source(source);
        // insta::assert_snapshot!(output);
    }
}
