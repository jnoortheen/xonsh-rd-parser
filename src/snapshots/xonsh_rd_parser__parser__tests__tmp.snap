---
source: src/parser.rs
expression: output
snapshot_kind: text
---
## AST

```
Module(
    ModModule {
        range: 0..18,
        body: [
            Expr(
                StmtExpr {
                    range: 0..18,
                    value: Call(
                        ExprCall {
                            range: 0..18,
                            func: Name(
                                ExprName {
                                    range: 0..5,
                                    id: Name("print"),
                                    ctx: Load,
                                },
                            ),
                            arguments: Arguments {
                                range: 5..18,
                                args: [
                                    Call(
                                        ExprCall {
                                            range: 7..17,
                                            func: Attribute(
                                                ExprAttribute {
                                                    range: 7..17,
                                                    value: Call(
                                                        ExprCall {
                                                            range: 7..17,
                                                            func: Attribute(
                                                                ExprAttribute {
                                                                    range: 7..10,
                                                                    value: Name(
                                                                        ExprName {
                                                                            range: 7..10,
                                                                            id: Name("__xonsh__"),
                                                                            ctx: Load,
                                                                        },
                                                                    ),
                                                                    attr: Identifier {
                                                                        id: Name("Pattern"),
                                                                        range: 7..10,
                                                                    },
                                                                    ctx: Load,
                                                                },
                                                            ),
                                                            arguments: Arguments {
                                                                range: 7..17,
                                                                args: [
                                                                    StringLiteral(
                                                                        ExprStringLiteral {
                                                                            range: 10..17,
                                                                            value: StringLiteralValue {
                                                                                inner: Single(
                                                                                    StringLiteral {
                                                                                        range: 10..17,
                                                                                        value: "hello",
                                                                                        flags: StringLiteralFlags {
                                                                                            quote_style: Single,
                                                                                            prefix: Regex,
                                                                                            triple_quoted: false,
                                                                                        },
                                                                                    },
                                                                                ),
                                                                            },
                                                                        },
                                                                    ),
                                                                ],
                                                                keywords: [],
                                                            },
                                                        },
                                                    ),
                                                    attr: Identifier {
                                                        id: Name("invoke"),
                                                        range: 7..17,
                                                    },
                                                    ctx: Load,
                                                },
                                            ),
                                            arguments: Arguments {
                                                range: 7..17,
                                                args: [
                                                    Name(
                                                        ExprName {
                                                            range: 7..10,
                                                            id: Name("foo"),
                                                            ctx: Load,
                                                        },
                                                    ),
                                                ],
                                                keywords: [],
                                            },
                                        },
                                    ),
                                ],
                                keywords: [],
                            },
                        },
                    ),
                },
            ),
        ],
    },
)
```
